### PR TITLE
Combine `UnsignedIntegerCoder` and `CFMaskCoder`

### PR DESCRIPTION
**Pull request by djhoese**
See #9266 and the related issues for the detailed discussion. Bottom line is that CF `_Unsigned` handling gets weird when handling `_FillValue`. The CF standard says that `_FillValue` on disk should always match the array's type on disk. However, when xarray loads the data and masks it, it does this with the in-memory unsigned integer data. Currently in `main` this is handled by a combination of the `UnsignedIntegerCoder` class and the `CFMaskCoder` class. It unfortunately requires the "decoded" unsigned `_FillValue` to be stored in the loaded variable so it can be used by the masking code. But to match CF standards this `_FillValue` should remain as the on-disk signed type. In this PR I combine the two classes to avoid storing the temporary unsigned version of `_FillValue` and only use it for masking.

Other important changes:

1. A serialization warning is added to inform users that the `_FillValue` they have passed does not match expectations for the CF standard.
2. `_FillValue` values that are numpy scalar types (ex. `np.uint8(255)`) are always converted to native python integers before being encoded for `_Unsigned` variables by calling `.item()` on them. This ensures the above serialization warning is always issued as numpy will silently cast a uint8 (ex. 255) to a int8 (ex. -1) without warning.

At the time of writing this code needs lots of cleanup or at least I hope it can because it is hard to follow in my opinion...but maybe that's just the way CF handling code is.

- [ ] Closes #xxxx
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/coding/variables.py [hunk 1]
```diff
@@ -349,46 +361,115 @@ def encode(self, variable: Variable, name: T_Name = None):
...
+            safe_setitem(encoding, attr, value, name=name)
+            for attr, value in raw_fill_dict.items()
+        ]
 
-                transform = partial(
-                    _apply_mask,
-                    encoded_fill_values=encoded_fill_values,
-                    decoded_fill_value=decoded_fill_value,
-                    dtype=dtype,
+        if "_Unsigned" in attrs:
```
Comment (dcherian) [inline xarray/coding/variables.py:417]: can we factor this out to its own function please

Comment (djhoese) [reply xarray/coding/variables.py:417]: That's why it is a draft and I said "At the time of writing this code needs lots of cleanup or at least I hope it can because it is hard to follow in my opinion".

Comment (djhoese) [reply xarray/coding/variables.py:452]: Very basic refactoring done for this specific if block. But it is leading me to some questions that I'll comment inline...

File: xarray/coding/variables.py [hunk 2]
```diff
@@ -298,18 +298,61 @@ def _check_fill_values(attrs, name, dtype):
...
+                # use view here to prevent OverflowError
+                encoded_fill_values.add(new_fill.view(unsigned_dtype).item())
+            data = lazy_elemwise_func(data, transform, unsigned_dtype)
+    elif data.dtype.kind == "u":
+        if unsigned == "false":
+            signed_dtype = np.dtype(f"i{data.dtype.itemsize}")
+            transform = partial(np.asarray, dtype=signed_dtype)
+            data = lazy_elemwise_func(data, transform, signed_dtype)
+            if raw_fill_value is not None:
+                new_fill = signed_dtype.type(raw_fill_value)
```
Comment (djhoese) [inline xarray/coding/variables.py:324]: This block didn't use @kmuehlbauer's trick for handling overflow by using `.view`. I'm wondering if I need to use that here, but no tests hit it. I've never actually seen `_Unsigned == "false"` in the wild.

Comment (djhoese) [reply xarray/coding/variables.py:324]: If I change that here then I think I might be able to shrink this function and do things as "old dtype" and "new dtype" rather than signed versus unsigned.

Comment (kmuehlbauer) [reply xarray/coding/variables.py:324]: > This block didn't use @kmuehlbauer's trick for handling overflow by using .view. I'm wondering if I need to use that here, but no tests hit it. I've never actually seen _Unsigned == \"false\" in the wild.

That was requested at some point in time for a specific use case. I'll try to dig it up.

Comment (djhoese) [reply xarray/coding/variables.py:324]: OK looking at the tests, it looks like this case is not tested (but I'm triple checking). This set of `if`s says that the data on-disk is unsigned, but `_Unsigned` is false which means the user wants signed data in-memory. The tests for `_Unsigned="false"` have signed data on-disk and in-memory so no casting/conversion happens.

Comment (kmuehlbauer) [reply xarray/coding/variables.py:324]: xref https://github.com/pydata/xarray/pull/4966

Comment (djhoese) [reply xarray/coding/variables.py:324]: Well I thought I was being smart and added tests for the `_Unsigned: "false"` case and now things are just all weird. It made me realize I wasn't handling this case in the encoding step, but it also kind of seems like it never was handled or that I have the wrong impression of how that configuration is supposed to be handled. The tests added in that PR @kmuehlbauer don't present the `_FillValue` so I've tried adding that to the backend tests but now non-NC4 backends are complaining about converting uint8 to int8. This coercion was added to solve https://github.com/pydata/xarray/issues/4014 it seems. 

My assumption is that if `_Unsigned: "false"` then the data is saved as uint8 and `_FillValue` should be uint8. But again, I'm not sure why my new tests are even trying to get to int8. I'll do some more testing later tonight hopefully. I'm not sure if it is better to spend a ton of time getting this small functionality working "as expected" or leave it undefined/untested as it is right now.

**Comment by djhoese** [comment]
I've run out of time to really work on refactoring this more. If people are able to review it in its current state that would be great. I'm not sure if it is in a good enough place to be merged, but I'll let the reviewers decide.

File: xarray/coding/variables.py (hunk 2 above)
Comment (djhoese) [reply xarray/coding/variables.py:324]: So my main concern and all of your handling in the decode pipeline for casting fill values turns out to not be an issue anymore as the raw fill values turn out to be numpy scalars (`np.uint8`) when they get loaded from the file. Or at least they are for the NetCDF4 cases. So numpy is perfectly happy casting uint8 to int8 and back if they are already numpy scalars. If my new tests cases make sense then I *think* this is fine.

**Comment by kmuehlbauer** [comment]
Thanks @djhoese for tackling this hard part of the CF conventions. This is looking good to me. Let's have another CI run now.

**Review by kmuehlbauer — APPROVED** [review]

**Comment by dcherian** [comment]
Kai, please merge if you're comfortable with this!

**Comment by kmuehlbauer** [comment]
Thanks @djhoese, for going down that rabbit hole. Finger's crossed :grimacing:.
